### PR TITLE
[Security] Fix user role restriction.

### LINKED ISSUE #35509: Can't login manually with multiple roles
Yesterday my project moved from 4.3.4 to 4.4.3 and I got a problem. 

I have a piece of code that allows to login as other user without sign out. So admin can login with ROLE_ADMIN first and then login as user and have roles: ROLE_ADMIN, ROLE_OWNER, ROLE_USER.  In result admin will have permissions of admin and user. He can view admin and user pages at the same time.

Listing: 
```
public function loginAsClient(TokenStorageInterface $tokenStorage, User $user)
{
    $user->setRoles(array_unique(array_merge($this->getUser()->getRoles(), $user->getRoles())));

    $token = new UsernamePasswordToken($user, null,  'main', $user->getRoles());
    $tokenStorage->setToken($token);

    $this->get('session')->set('_security_main', serialize($token));
    $this->get('session')->save();

    return $this->redirectToRoute('dashboard_index');
}
```

And everything worked until I did upgrade. Now nothing happens after this action. Just redirection without saving of new roles in a session.

Also, I notices that if do not merge roles, and give just user roles(without ROLE_ADMIN) authentication will work, but now admin will lost admin permissions and logout will be required. 

Has something changed in the new security version or is this type of manual login is deprecated?

### PR DESCRIPTION
**Pull request by gorshkov-ag**
It was impossible to restrict user roles on the fly without deauthentication.

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #35509
| License       | MIT
| Doc PR        |

Allowed to restrict current user roles without changing roles for user object.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Can you add a test case please?

**Comment by xabbuh** [comment]
> It was impossible to restrict user roles on the fly

@gorshkov-ag Can you explain a bit what you mean with that?

**Comment by gorshkov-ag** [comment]
> > It was impossible to restrict user roles on the fly
> 
> @gorshkov-ag Can you explain a bit what you mean with that?

I have some user roles in application and need a possibility to switch them without changing user or user roles.

**Comment by gorshkov-ag** [comment]
> Can you add a test case please?

Added test cases. Updated rule don't compare count of current user roles with count of applicable roles.

**Comment by nicolas-grekas** [comment]
@oleg-andreyev could you please have a look, since this changes the logic from #31177?

File: src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php [hunk 1]
```diff
@@ -276,7 +276,7 @@ private function hasUserChanged(UserInterface $user): bool
             $userRoles[] = 'ROLE_PREVIOUS_ADMIN';
         }
 
-        if (\count($userRoles) !== \count($this->getRoleNames()) || \count($userRoles) !== \count(array_intersect($userRoles, $this->getRoleNames()))) {
+        if (\count($this->getRoleNames()) !== \count(array_intersect($this->getRoleNames(), $userRoles))) {
```
Comment (oleg-andreyev) [inline src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: Can you please update also https://github.com/symfony/symfony/blob/4f4c30d59ed429bc08a34f330c45de2b04059d0a/src/Symfony/Component/Security/Core/User/User.php#L148

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php [hunk 2]
```diff
@@ -72,5 +72,53 @@ public function getSalt()
         $token = new SwitchUserToken($impersonated, 'bar', 'provider-key', ['ROLE_USER', 'ROLE_PREVIOUS_ADMIN'], $originalToken);
         $token->setUser($impersonated);
         $this->assertTrue($token->isAuthenticated());
+
+        $impersonator = new class() implements UserInterface {
```
Comment (oleg-andreyev) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php]: I think it's better to create a separate unit for this

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php [hunk 3]
```diff
@@ -72,5 +72,53 @@ public function getSalt()
...
+        $token->setUser($impersonator);
+        $this->assertTrue($token->isAuthenticated());
+
+        $token = new SwitchUserToken($impersonator, 'foo', 'provider-key', ['ROLE_ADMIN', 'ROLE_PREVIOUS_ADMIN'], $originalToken);
+        $token->setUser($impersonator);
+        $this->assertTrue($token->isAuthenticated());
+
+        $token = new SwitchUserToken($impersonator, 'foo', 'provider-key', ['ROLE_TEST', 'ROLE_PREVIOUS_ADMIN'], $originalToken);
+        $token->setUser($impersonator);
+        $this->assertFalse($token->isAuthenticated());
```
Comment (oleg-andreyev) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php]: `assertFalse` contradicts with the unit name `testSetUserDoesNotDeauthenticate`

**Review by oleg-andreyev — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php (hunk 1 above)
Comment (gorshkov-ag) [reply src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: Done https://github.com/gorshkov-ag/symfony/commit/429dee7efb5d018a67d0654bfbc956cb08d1b86c

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php (hunk 3 above)
Comment (gorshkov-ag) [reply src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php]: Updated test cases https://github.com/gorshkov-ag/symfony/commit/9de09f0cfbfed2af1f7ae0a63597171615a51ac3

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php (hunk 2 above)
Comment (gorshkov-ag) [reply src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php]: Now is it problem in User equals test: User without role now is equivalent with user with role "ROLE", but opposite is not true.
https://github.com/gorshkov-ag/symfony/blob/9de09f0cfbfed2af1f7ae0a63597171615a51ac3/src/Symfony/Component/Security/Core/Tests/User/UserTest.php#L110-L120
120 line case.

File: src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php (hunk 1 above)
Comment (gorshkov-ag) [reply src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: Now is it problem in User equals test: User without role now is equivalent with user with role "ROLE", but opposite is not true.
https://github.com/gorshkov-ag/symfony/blob/9de09f0cfbfed2af1f7ae0a63597171615a51ac3/src/Symfony/Component/Security/Core/Tests/User/UserTest.php#L110-L120
120 line case.

Comment (dmaicher) [inline src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: hmm simply comparing the count of roles is not really enough, or? What if I remove one role and add another one? Then the count is still the same but the roles have changed?

**Comment by chalasr** [comment]
The current behavior is the expected one.
Looking at the linked issue, it seems like you are manually changing the session token from a controller, which is probably not a good idea.
👎 for this change.

**Comment by gorshkov-ag** [comment]
Not removing role for user, just for token to use one or some user roles, not all of them.

**Comment by chalasr** [comment]
The token attributes must reflect the roles of the user it holds.
The only exception to this rule is `ROLE_PREVIOUS_ADMIN`(impersonation) as it is added on the fly, the logic for this case is hardcoded in AbstractToken.

There is no valid use case for having a token that knows only part of the user roles, as it does not reflect the real state of the user.

**Comment by dmaicher** [comment]
I think this default behavior of de-authenticating on any role change seems good. So not sure we should change anything here as @chalasr said.

@gorshkov-ag maybe for your usecase its enough to make your user implement `EquatableInterface` and customize the logic when to consider a user changed? This would by-pass the default behavior.

**Comment by fabpot** [comment]
Closing as I agree with @chalasr

**Comment by jmalinens** [comment]
@fabpot this has bitten us after symfony4.4 upgrade because by default we give ROLE_USER role but when user accesses admin panel (separate symfony instance which shares session with main website) and he is in the list of admins we add ROLE_ADMIN for him.

After symfony upgrade this does not work anymore as adding ROLE_ADMIN deauthenticates user.

This is one of the multiple breaking changes in symfony authentication in minor versions we have to deal with :/

**Comment by dmaicher** [comment]
@jmalinens its hard (impossible) to anticipate every possible way people are using symfony and rely on behavior that was never intended :wink: 

So see the positive side here: now this makes things more secure by default. If you need to restore the previous behavior its quite simple: make your user implement `EquatableInterface` and do not check the roles at all.

**Comment by oleg-andreyev** [comment]
@jmalinens I'm not aware of your implementation but you need to make sure that you ADD not REPLACE your ROLE_USER.

_Probably_ `\count($currentRoles) !== \count($newRoles)` is too strict and here is an example that @gorshkov-ag and @jmalinens faced:
http://sandbox.onlinephpfunctions.com/code/0bedfc6e52a54307fa06c91154a9854a81d0cecc

**Comment by jmalinens** [comment]
Thanks guys.
I went with implementing `EquatableInterface` and at the same time got rid of AdvancedUserInterface.

**Pull request closed without merging**